controller/forwarding/hyperlane_test.go

```diff
@@ -0,0 +1,475 @@
+
+	testCases := []struct {
+		name         string
+		setup        func(*mocks.HyperlaneHandler)
+		hypAttr      *forwardingtypes.HypAttributes
+		transferAttr *types.TransferAttributes
+		expError     string
+	}{
+		{
+			name: "success - when the attributes are valid",
+			setup: func(m *mocks.HyperlaneHandler) {
+				tokenID := hyperlaneutil.HexAddress(usdnID)
+				hypToken := &warptypes.WrappedHypToken{
+					Id:            tokenID.String(),
+					Owner:         "noble",
+					TokenType:     0,
+					OriginMailbox: "",
+					OriginDenom:   "usdn",
+					IsmId:         &hyperlaneutil.HexAddress{},
+				}
+
+				m.Tokens[hypToken.Id] = *hypToken
+			},
+			transferAttr: transferAttr,
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
+		},
+		{
+			name: "error - when the hyperlane denom is not destination denom",
+			setup: func(m *mocks.HyperlaneHandler) {
+				tokenID := hyperlaneutil.HexAddress(usdnID)
+				hypToken := &warptypes.WrappedHypToken{
+					Id:            tokenID.String(),
+					Owner:         "noble",
+					TokenType:     0,
+					OriginMailbox: "",
+					OriginDenom:   "btc",
+					IsmId:         &hyperlaneutil.HexAddress{},
+				}
+
+				m.Tokens[hypToken.Id] = *hypToken
+			},
+			transferAttr: transferAttr,
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
+			expError: "invalid forwarding token",
+		},
+		{
+			name:         "error - when hyperlane attributes are nil",
+			setup:        func(m *mocks.HyperlaneHandler) {},
+			transferAttr: transferAttr,
+			hypAttr:      nil,
+			expError:     "invalid Hyperlane attributes",
+		},
+		{
+			name:         "error - when transfer attributes are nil",
+			setup:        func(m *mocks.HyperlaneHandler) {},
+			transferAttr: nil,
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
+			expError: "invalid transfer attributes",
+		},
+		{
+			name:         "error - when the token does not exist",
+			setup:        func(m *mocks.HyperlaneHandler) {},
+			transferAttr: transferAttr,
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
```

Comment:
I think it would be nice to refactor this because currently there is a big amount of duplication and while reviewing it was repetitive to check each time if some individual field has changed. Rather we can store this e.g. as `validHypAttributes` and reuse for the different tests.